### PR TITLE
feat(automation): add a keyevent verb for momentary shortcuts (#5079)

### LINKED ISSUE #5079: Automation bridge: momentary (event-filter-driven) shortcuts cannot be exercised — no verb delivers a key press/release edge
## What

The app has two kinds of keyboard action, and the bridge can only reach one.
Shortcuts with a direct handler fire via the `shortcut` verb; the momentary
family — PTT hold and the CW momentary keys — lives in the application event
filter as KeyPress/KeyRelease *edges* with their own state bookkeeping
(`src/gui/MainWindow_Shortcuts.cpp:134` `handleCwMomentaryShortcut`, `:209`
`handlePttHoldShortcut`, `m_pttHoldActive`). The bridge's own error message
states the gap: firing one returns *"exists but is event-filter-driven
(momentary key action) — it has no direct handler the bridge can fire"*
(`src/core/AutomationServer.cpp:8118-8122`).

`key ptt` doesn't cover it either, by design: it routes to
`RadioModel::setTransmit` rather than synthesizing a QKeyEvent
(`src/core/AutomationServer.cpp:7781-7810`) — the right default for TX
control, but it bypasses the event-filter path entirely, so the momentary
handlers' flag bookkeeping never runs. No verb constructs a key event
anywhere in `AutomationServer.cpp` (the only two `QKeyEvent` mentions are the
comments above).

## What this hides

Any bug that lives in the momentary handlers' press/release state machine
cannot be reproduced by the bridge. Measured 2026-08-06 (Linux, driving a
repro for #4638) [RIG]: the PTT-hold path could not be exercised at all —
external XTEST key injection was defeated by the window manager's
focus-stealing prevention, and the proof fell back to operator hands.

## Suggested shape

A `keyevent <press|release> <key>` verb posting a real QKeyEvent through the
application (or an invokable that drives the two handlers with synthetic
events). Since these edges key RF, it must sit behind the same TX guards as
the existing `key`/`shortcut` TX paths (`ShortcutFireTxBlocked` /
`AETHER_AUTOMATION_ALLOW_TX`) — a press with no matching release should also
be covered by the existing fail-safe (`failSafeMomentaryKeyingToRx`).
Whether and how is the maintainer's call.

## Non-goals

No change to `key ptt`/`key mox` — model routing is the right default for
plain TX control. Not a general keyboard-macro facility; the target is the
momentary shortcut family the event filter owns.

## Relations

Same species as #5068 and #5078: a single missing input primitive blocking
an otherwise fully drivable verification workflow.

— authored by agent (Claude Code) on behalf of @skerker

### PR DESCRIPTION
**Pull request by skerker**
## Summary

The momentary family — PTT hold and the CW straight/dit/dah keys — lives in the application
event filter as KeyPress/KeyRelease *edges* with its own state bookkeeping
(`handleCwMomentaryShortcut`, `handlePttHoldShortcut`), so `shortcut` cannot fire it
(`ShortcutFireNoDirectHandler`) and `key ptt` bypasses it through the model. No verb constructed
a `QKeyEvent`. Per the review on #5079, with both of its design deltas:

- **`keyevent <press|release> <action-id|key-seq>`** (`doKeyEvent` →
  `MainWindow::injectKeyEventForAutomation`, Q_INVOKABLE). An action id resolves to its
  *current* binding through `ShortcutManager` first, so a rebind moves the injected key (#3879);
  a literal sequence such as `Ctrl+T` is the escape hatch for the modifier-tolerant release
  branch. Unrecognised specs are rejected — including `QKeySequence::fromString`'s non-empty
  `Key_unknown` parse, which the pre-commit smoke caught slipping through an `isEmpty()` check.
- **`sendEvent`, not `postEvent`**, to the main window itself — never the focus widget (second
  commit, per adversarial review): the qApp-level filter runs synchronously for any receiver and
  the momentary handlers never look at it, so the reply still carries `consumed`; a synthesized
  edge that only the filter sees cannot click a widget that keys TX on its own (a focused
  ATU/CWX/APRS button on Space, a dialog's default button on Return) — the `aetherTxKeying`
  marker `invoke` honours is never bypassed (Principle VI). No window manager is involved, which
  is what defeated the XTEST attempt in the issue. `consumed` means no handler called
  `ignore()`, so a swallow-all branch (SWR sweep) also reports `true`; which handler claimed the
  edge is not observable from the reply.
- **TX gate on the press only.** A `keysTx` press is refused unless `AETHER_AUTOMATION_ALLOW_TX`
  is set (`KeyInjectTxBlocked`, same wording as the `key`/`shortcut` guards); a release is always
  delivered, since refusing it would strand a keyed transmitter — the failure
  `failSafeMomentaryKeyingToRx` exists to prevent. Result codes: `Ok` / `UnknownKey` /
  `TxBlocked` / `NotConsumed` / `TxOk` / `Unbound` (a known action id with no key binding —
  the CW momentary ids ship unbound — answers that, not "not a known action id").
- **Watchdog ownership** follows the review's tighter option: `markTxBridgeInitiated()` only on
  `TxOk` (a keysTx press the filter actually claimed); a release hands policing back **only when
  the transmitter is actually down** (third commit, per review): a release the momentary handler
  declined — text entry focused, radio dropped — or one that un-keyed nothing because TX is up
  from another source leaves the flag armed, so a leaked press keeps its backstop
  (`onTxWatchdog()` clears the flag itself on the next poll that finds nothing keyed).
  `automation_tx_watchdog_test` pins it: release while still keyed → still armed; release with TX
  down → cleared; fails against the unconditional clear (checked).
  As the review notes, `failSafeMomentaryKeyingToRx` fires on deactivation, so in a headless
  bridge session the TX watchdog (`m_txMaxKeyMs` / `forceUnkey()`) is the backstop for a leaked
  press — stated in the code comment at the arm site.

`docs/automation-bridge.md` regenerated (`tools/gen_bridge_docs.py`, `--check` clean). `key ptt`
/ `key mox` unchanged. `keyevent` is a new bridge verb (documented surface) — the maintainer's call
whether the verb is the right shape. In observe-only bridge mode it is refused like every
non-allow-listed verb. A caller should not have a text entry focused: the event goes to the focus
widget, and the momentary handlers then decline it (`consumed: false`, reported, not an error).
No unit test: nothing under `tests/` can reach `MainWindow`; the `Key_unknown` parse rule is the
one piece a pure-Qt test could pin and is a candidate follow-up.

## Constitution principle honored

Principle VI — TX safety: the new input path keys nothing without the explicit opt-in, never
blocks the un-keying edge, and hands a keyed press to the existing watchdog.

## Test plan

- [x] Local build passes — clean configure + full build, macOS (Intel, Qt 6.8.3): `f84a0f6e`, then `c01488f2` (embedded SHA verified in the binary before the run)
- [x] Behavior verified on a real radio if applicable — Demo radio via the agent automation bridge, `AETHER_AUTOMATION_ALLOW_TX` unset (no TX possible, none attempted): readings in **Proof**
- [x] Existing tests pass (CI) — `automation_tx_watchdog_test` passes locally on `8e5c3da0` with the new case (not on a `ci.yml` `-R` gate, like the other `automation_*` tests); `tools/gen_bridge_docs.py --check` clean
- [x] Reproduction steps documented — before: `shortcut ptt_hold` → "exists but is event-filter-driven…"; after: see Proof

### Proof

Readings on the clean build at `c01488f2` (the main-window-delivery revision; `whoami` reports
`txAllowed: false`), Demo radio connected, `AETHER_AUTOMATION_ALLOW_TX` unset; `transmitting: false`
throughout. The first round on `f84a0f6e` read identically for the five original rows.

| call | reading |
|---|---|
| `keyevent bogus ptt_hold` | `{"ok":false,"error":"keyevent needs press\|release"}` |
| `keyevent press no_such_action_xyz` | `{"ok":false,"error":"not a known action id or key sequence: no_such_action_xyz"}` |
| `keyevent press ptt_hold` | `{"ok":false,"error":"blocked: 'ptt_hold' keys the transmitter (TX-safety guard). Set AETHER_AUTOMATION_ALLOW_TX=1 to override."}` |
| `keyevent release ptt_hold` | `{"ok":true,"keyevent":"release","key":"ptt_hold","consumed":true}` — delivered un-gated and claimed by the app event filter (`consumed` = `QKeyEvent::isAccepted()`); which handler claimed it is not observable from the reply |
| `keyevent release Ctrl+T` | `{"ok":true,"keyevent":"release","key":"Ctrl+T","consumed":false}` — literal sequence delivered, no momentary handler bound to it |
| `keyevent press cwkey` (the CW straight key, registered unbound) | `{"ok":false,"error":"action 'cwkey' exists but has no key binding; bind it in Settings or pass a literal key sequence"}` — the new `Unbound` answer, no longer "not a known action id" |

Before: `shortcut ptt_hold` answers *"exists but is event-filter-driven (momentary key action) — it
has no direct handler the bridge can fire"* (quoted in #5079; on this build the same call is stopped
earlier by the TX guard, so the quote is the issue's measurement). The keyed-press arm (`TxOk`,
watchdog armed) needs `AETHER_AUTOMATION_ALLOW_TX=1` and a dummy load and was not run.

## Checklist

- [x] Commits are signed (3, SSH)
- [x] No new flat-key `AppSettings` calls — **n/a**
- [x] Code is clean-room
- [x] All meter UI uses `MeterSmoother` — **n/a**
- [x] Documentation updated if user-visible behavior changed — `docs/automation-bridge.md` regenerated
- [x] Security-sensitive changes reference a GHSA — **n/a**; TX-gated like the existing keying verbs

Fixes #5079

— authored by agent (Claude Code) on behalf of @skerker

### CODE REVIEW

**Comment by skerker** [comment]
@AetherClaude please re-review

**Comment by skerker** [comment]
Review round addressed. The blocker landed earlier in `8e5c3da0` (release hands policing back only via `txBridgeOwnsCurrentTransmit()`, watchdog-test case fails against the unconditional clear). In `69274478`: multi-chord sequences are rejected instead of silently truncated; the verb help now says the CW momentary ids ship unbound (bind in Configure Shortcuts first) and carries the Tab/Backtab consumed caveat; the KeyInjectTxOk arming comment acknowledges the over-arming window and why it self-heals. Branch is merged with current `main` (`73738c89`); docs regenerated, `--check` clean; `automation_tx_watchdog_test` and `automation_fm_repeater_verbs_test` pass on the merged tree.

— authored by agent (Claude Code) on behalf of @skerker

File: src/gui/MainWindow_Shortcuts.cpp [hunk 1]
```diff
@@ -1300,6 +1303,65 @@ int MainWindow::fireShortcutAction(const QString& id, bool allowTx)
...
+    if (seq.count() > 1)
+        return KeyInjectUnknownKey;
+
+    // TX gate on the PRESS only. Blocking a release would leave the transmitter
+    // keyed with no way to drop it — the failure failSafeMomentaryKeyingToRx
+    // exists to prevent. A release is always safe to deliver.
+    const bool keysTx = a && a->keysTx;
+    if (press && keysTx && !allowTx)
+        return KeyInjectTxBlocked;
+
```
Comment (ten9876) [inline src/gui/MainWindow_Shortcuts.cpp:1342]: This is the right way round, and I want to say so explicitly because the opposite is the tempting mistake.

> TX gate on the PRESS only. Blocking a release would leave the transmitter keyed with no way to drop it

A reviewer pattern-matching on "TX verb, add a safety check" would gate both edges, and that produces a transmitter that cannot be unkeyed through the same channel that keyed it. Gating the press and always delivering the release is the fail-safe reading of Principle VI, not the lenient one.

I did check the one way this could be hollow: `const bool keysTx = a && a->keysTx;` means a literal key sequence whose `actionForKey()` lookup finds nothing gets `keysTx == false` and skips the gate entirely. For this family it cannot happen — every momentary action is registered with `keysTx=true` at `:851`, `:859`, `:864`, `:872`, `:877` and `:1203`, and PTT hold resolves through `kPttHoldActionId` rather than a hardcoded `Key_Space`, so a rebind moves it and `actionForKey()` still finds it.

The residual exposure is a future TX-keying key handled purely in the event filter with no `ShortcutManager` registration — it would inject un-gated. Nothing like that exists today, and the `keysTx` flag's own comment already tells registrants what it is for, so this is a note rather than a request.

**The nit:** two consecutive `press` edges with no intervening `release`. The watchdog means nothing runs away, and `releaseEdgeHandsBackPolicing()`'s fail-closed check means a spurious release cannot disarm it — I mutation-tested that one and the suite catches it. But a driver that drops a release and re-presses is the realistic way this verb gets misused, and neither the code nor `docs/automation-bridge.md` says whether the second press re-arms the watchdog, extends it, or is a no-op. One sentence wherever the press branch arms it would close that.

**Review by ten9876 — COMMENTED** [review]
## Issue fit

#5079 asks for a bridge verb that can drive the momentary family — PTT hold and the CW straight/dit/dah keys — which live in the qApp event filter as press/release *edges* with their own bookkeeping, so `shortcut` cannot reach them and `key ptt` bypasses them through the model. A synthesized `QKeyEvent` is the only thing that exercises the real path, and this builds one carefully.

**Verified empirically** on macOS/arm64: the full `AetherSDR` target builds clean and `automation_tx_watchdog_test` passes (4.6 s). I also mutation-tested the safety-critical invariant — making the release edge hand back policing unconditionally:

```
[FAIL] a release edge while the transmitter is still keyed does not disarm the watchdog
```

So the fail-closed behaviour is genuinely pinned, not just commented.

**No blockers.** This is the most careful TX-adjacent change I have reviewed in this batch, and the reasoning is worth naming rather than waving through.

### Principle VI, done properly

- **The gate is on the press only**, with the right justification: *"Blocking a release would leave the transmitter keyed with no way to drop it."* Refusing a release would be the safety-shaped-looking choice and the wrong one; getting this the right way round is the whole ballgame for a momentary verb.
- **The gate has something to gate on.** I checked whether `a->keysTx` can be null for a literal key sequence — `actionForKey()` returning nothing would silently skip the TX check. It cannot, for this family: every momentary action is registered with `keysTx=true` (`MainWindow_Shortcuts.cpp:851, 859, 864, 872, 877, 1203`), including the ones that ship unbound, and PTT hold resolves through `kPttHoldActionId` rather than a hardcoded `Key_Space`. So both the id path and the literal-sequence path land on the same flag.
- **`releaseEdgeHandsBackPolicing()` fails closed** — if the bridge still owns the transmit, the watchdog stays armed rather than being disarmed by a release that did not actually unkey.
- **`sendEvent` to the main window, never the focus widget.** The reasoning in the second commit is exactly right: a synthesized edge delivered to a focused ATU/CWX/APRS button on Space could click a control that keys TX on its own, and the qApp filter runs for any receiver anyway, so nothing is lost by not targeting focus.

Two Qt traps also handled that I would expect most PRs to miss: `QKeySequence::fromString` parsing garbage into a *non-empty* sequence whose key is `Key_unknown` (so `isEmpty()` alone lets it through), and multi-chord sequences being silently truncated to their first chord by `seq[0]` — rejected rather than half-delivered, so the verb never reports `consumed` for something it did not send. Distinguishing "unbound action id" from "unknown action id" in the error is a small thing that will save someone real time.

## Scope

Everything is explained by #5079. No `CHANGELOG.md` entry — correct. `docs/automation-bridge.md` and the header verb list are both updated, and `69274478` walks back the verb's documented limits to what it actually does, which is the right direction to revise in.

## Nits

- *(inline: `MainWindow_Shortcuts.cpp:1340`)* Worth a line on what two consecutive `press` edges with no intervening `release` do. The watchdog is the backstop so nothing runs away, but a driver that loses a release and re-presses is the realistic failure mode for a momentary verb, and the code does not say whether that re-arms, extends, or is idempotent.
- `injectKeyEventForAutomation()` returns bare `int` codes (`KeyInjectUnbound`, `KeyInjectUnknownKey`, `KeyInjectTxBlocked`). Pragmatic for `Q_INVOKABLE` and clearly named — noting only that the values cross the `invokeMethod` boundary untyped, so a future code added on one side and not handled on the other fails silently as "some other int".

## Worth surfacing beyond this PR

`cf595faf` is *"ci: retrigger after `icom_backend_test` flake on the Linux gate"* — the **second** such commit I have hit today; #5144 carries `86f7ea20` with the same message. Two contributors independently burning a commit to get past the same flake, on a test that gates every Icom change, is a pattern rather than bad luck. I ran `icom_backend_test` about a dozen times across this batch and it passed every time on macOS, at 30–97 s depending on branch — against the *"~21 s for the pair"* its own gate comment claims — so I cannot reproduce it here, but the timing drift alone suggests that comment is stale. Probably worth an issue with links to both runs.

## Verified vs. read

**Built and ran:** the full app target, `automation_tx_watchdog_test`, and the fail-open mutation above (reverted). **Traced:** every `keysTx=true` registration in the momentary family, to confirm the TX gate cannot be bypassed via a literal key sequence. **Read:** the `sendEvent`-vs-`postEvent` and main-window-vs-focus reasoning, and the `Key_unknown` / multi-chord guards. **Not verified:** no radio here, so nothing was actually keyed — the watchdog behaviour is exercised by the test's own harness rather than against hardware.

**Review by jensenpat — APPROVED** [review]
## Issue fit

**Yes.** #5079 asks for a bridge primitive that delivers real press/release edges through the existing application event filter so PTT-hold and bound CW momentary shortcuts can be exercised without bypassing their state bookkeeping. The current head adds exactly that path, keeps presses behind `AETHER_AUTOMATION_ALLOW_TX`, always permits release, delivers only to the main window, and retains watchdog ownership until TX is actually down. The earlier watchdog-disarm blocker is fixed and pinned by the added regression test. Project maintainer @ten9876 has already reviewed the public `keyevent` shape and reported no blockers.

## Scope

| File/group | Change | Claimed by issue/body? | Verdict |
|---|---|---|---|
| `src/core/AutomationServer.{h,cpp}` | Registers and dispatches `keyevent`; applies TX gating and watchdog hand-back | Yes | In scope |
| `src/gui/MainWindow.h`, `MainWindow_Shortcuts.cpp` | Resolves current bindings and synchronously injects a single key edge through the existing filter | Yes | In scope |
| `tests/automation_tx_watchdog_test.cpp` | Pins fail-closed release/watchdog ownership | Yes | In scope |
| `docs/automation-bridge.md` | Regenerated verb documentation, including unbound-CW and Tab caveats | Yes | In scope |

Everything in the diff is explained by #5079. There is no `CHANGELOG.md` entry, settings/schema change, new default binding, radio-backend change, unrelated UI preference, or removed guard. The PR body's scope claims hold.

## Blockers

None.

## Nits

None requiring a change. The open repeated-press question is safe in the current state machine: while PTT hold is already active the second edge performs no additional radio action, the pre-dispatch keyed sample prevents `markTxBridgeInitiated()` from resetting the existing deadline, and the original watchdog ownership remains armed.

## Verification

Verified against exact head `cf595faf` merged synthetically with current `origin/main` `cec77832`: clean `git merge-tree`, full macOS app build with `-j22`, headless socket-free `automation_tx_watchdog_test` 1/1, and `tools/gen_bridge_docs.py --check`. All six PR commits are GitHub-verified. The unrelated socket-based tests were explicitly excluded by operator direction. The first focused test launch exposed only a stale host Homebrew ONNX/protobuf dylib; rebuilding the test target with optional ONNX disabled allowed the test logic to run and pass.

No physical-radio TX was attempted; this review establishes code, build, and watchdog-test behavior, not dummy-load hardware proof.
